### PR TITLE
8246797: A convenient method to read OPTIONAL element

### PR DESCRIPTION
**Pull request by wangweij**
Add some DerInputStream methods to simplify OPTIONAL element reading. Update
the x509 classes to make use of the new methods.

---------
### Progress
- [x] Change must not contain extraneous whitespace

### Issue
 * [JDK-8246797](https://bugs.openjdk.java.net/browse/JDK-8246797): A convenient method to read OPTIONAL element

### Download
`$ git fetch https://git.openjdk.java.net/playground pull/11/head:pull/11`
`$ git checkout pull/11`

### CODE REVIEW

File: src/java.base/share/classes/sun/security/util/DerInputBuffer.java [hunk 1]
```diff
@@ -61,13 +61,12 @@
...
 
     DerInputBuffer dup() {
-        try {
-            DerInputBuffer retval = (DerInputBuffer)clone();
-            retval.mark(Integer.MAX_VALUE);
-            return retval;
-        } catch (CloneNotSupportedException e) {
-            throw new IllegalArgumentException(e.toString());
-        }
+        DerInputBuffer retval = new DerInputBuffer(this.buf);
```
Comment (valeriepeng) [inline src/java.base/share/classes/sun/security/util/DerInputBuffer.java:64]: This is same as calling clone()? If so, not seeing the benefit of having dup()...

File: src/java.base/share/classes/sun/security/util/DerInputBuffer.java [hunk 2]
```diff
@@ -61,13 +61,12 @@
...
+        retval.pos = this.pos;
+        retval.count = this.count;
+        retval.allowBER = this.allowBER;
+        retval.mark = this.pos;
+        return retval;
```
Comment (valeriepeng) [inline src/java.base/share/classes/sun/security/util/DerInputBuffer.java:69]: This is same as calling clone()? If so, not seeing the benefit of having dup()...

File: src/java.base/share/classes/sun/security/util/DerInputStream.java [hunk 3]
```diff
@@ -661,4 +684,118 @@ static int getDefiniteLength(InputStream in) throws IOException {
...
+    /**
+     * Detect if the tag of the next DerValue in the stream matches the rule.
+     *
+     * Attention: tag is an integer casted from a byte. i.e. could be negative.
+     *
+     * @param rule the rule to check for the tag.
+     * @return true if matches, false otherwise or stream is at end.
+     * @throws IOException if an I/O error happens
+     */
+    public boolean seeOptional(Predicate<Integer> rule) throws IOException {
```
Comment (valeriepeng) [inline src/java.base/share/classes/sun/security/util/DerInputStream.java:707]: rule may be too generic? Probably something like tagChecker?

File: src/java.base/share/classes/sun/security/util/DerInputBuffer.java (hunk 2 above)
Comment (wangweij) [reply src/java.base/share/classes/sun/security/util/DerInputBuffer.java:69]: There is an extra `mark` call but seems not used anywhere.

File: src/java.base/share/classes/sun/security/util/DerInputStream.java (hunk 3 above)
Comment (wangweij) [reply src/java.base/share/classes/sun/security/util/DerInputStream.java:707]: There are other methods based on this one.

File: src/java.base/share/classes/sun/security/util/DerValue.java [hunk 4]
```diff
@@ -945,4 +972,78 @@ public void resetTag(byte tag) {
     public int hashCode() {
         return toString().hashCode();
     }
+
+    /**
+     * Dump the content of this DerValue into stdout, decomposing the
+     * structire in a hierachical style.
```
Comment (valeriepeng) [inline src/java.base/share/classes/sun/security/util/DerValue.java:978]: typo: structure

File: src/java.base/share/classes/java/security/cert/PolicyQualifierInfo.java [hunk 5]
```diff
@@ -89,6 +91,28 @@
...
+            return;
+        }
+        DerOutputStream out = new DerOutputStream();
+        DerOutputStream ins = new DerOutputStream();
+        ins.putOID(ObjectIdentifier.of(mId));
+        ins.write(mData);
+        out.write(DerValue.tag_Sequence, ins);
+        mEncoded = out.toByteArray();
+    }
+
```
Comment (valeriepeng) [inline src/java.base/share/classes/java/security/cert/PolicyQualifierInfo.java:115]: This seem irrelevant to the RFE? It seems that existing impl only supports parsing? Are we really using this new constructor somewhere in JDK?

File: src/java.base/share/classes/sun/security/pkcs/PKCS7.java [hunk 6]
```diff
@@ -266,6 +266,14 @@ private void parseNetscapeCertChain(DerValue val)
...
     }
 
+
+    //    SignedData ::= SEQUENCE {
+    //        version CMSVersion,
+    //        digestAlgorithms DigestAlgorithmIdentifiers,
+    //        encapContentInfo EncapsulatedContentInfo,
+    //        certificates [0] IMPLICIT CertificateSet OPTIONAL,
+    //        crls [1] IMPLICIT RevocationInfoChoices OPTIONAL,
+    //        signerInfos SignerInfos }
```
Comment (valeriepeng) [inline src/java.base/share/classes/sun/security/pkcs/PKCS7.java:276]: We should include RFC 5652 here as the file is named PKCS7 which is in RFC 2315 which has different field names and types.

File: src/java.base/share/classes/sun/security/util/DerInputStream.java [hunk 7]
```diff
@@ -661,4 +684,118 @@ static int getDefiniteLength(InputStream in) throws IOException {
...
+
+    /**
+     * Detect if the next DerValue in the stream is a context-specific value
+     * tagged by {@code n}.
+     *
+     * @param n the expected tag
+     * @return true if is, false otherwise or stream is at end.
+     * @throws IOException if an I/O error happens
+     */
+    public boolean seeOptionalContextSpecific(int n) throws IOException {
```
Comment (valeriepeng) [inline src/java.base/share/classes/sun/security/util/DerInputStream.java:770]: Consider moving this seeOptionalXXX upward so that all seeXXX methods are together for easier reference.

File: src/java.base/share/classes/java/security/cert/PolicyQualifierInfo.java (hunk 5 above)
Comment (wangweij) [reply src/java.base/share/classes/java/security/cert/PolicyQualifierInfo.java:115]: I uses the constructor in a test. Yes it should not be included.

File: src/java.base/share/classes/sun/security/util/DerInputStream.java (hunk 7 above)
Comment (wangweij) [reply src/java.base/share/classes/sun/security/util/DerInputStream.java:770]: Good suggestion.

File: src/java.base/share/classes/sun/security/pkcs/PKCS7.java [hunk 8]
```diff
@@ -305,7 +313,7 @@ private void parseSignedData(DerValue val)
          * check if certificates (implicit tag) are provided
          * (certificates are OPTIONAL)
          */
-        if ((byte)(dis.peekByte()) == (byte)0xA0) {
+        if (dis.seeOptionalContextSpecific(0)) {
```
Comment (valeriepeng) [inline src/java.base/share/classes/sun/security/pkcs/PKCS7.java:316]: The new call did not check bit 6. We should either enhance the seeOptionalXXX method or add another one to ensure that bit 6 has the expected value as well.

Comment (wangweij) [reply src/java.base/share/classes/sun/security/pkcs/PKCS7.java:316]: You mean the constructed bit? The getSet() call in this if block will fail if it's not set. I wonder what difference it will make here. Suppose the bit 6 is not set but it's still [0], that means someone has stuffed a non-constructed data type here. In fact, since it's still [0], I think it should be better parsed inside this block and not somewhere else.

File: src/java.base/share/classes/sun/security/pkcs/SignerInfo.java [hunk 9]
```diff
@@ -129,6 +129,15 @@ public SignerInfo(DerInputStream derin)
...
      *
+     * SignerInfo ::= SEQUENCE {
+     *         version CMSVersion,
+     *         sid SignerIdentifier,
+     *         digestAlgorithm DigestAlgorithmIdentifier,
+     *         signedAttrs [0] IMPLICIT SignedAttributes OPTIONAL,
+     *         signatureAlgorithm SignatureAlgorithmIdentifier,
+     *         signature SignatureValue,
+     *         unsignedAttrs [1] IMPLICIT UnsignedAttributes OPTIONAL }
+     *
```
Comment (valeriepeng) [inline src/java.base/share/classes/sun/security/pkcs/SignerInfo.java:140]: Similar to earlier comment, existing variable naming matches better with the ASN.1 definition in RFC 2315, e.g. SignerInfo ::= SEQUENCE {
     version Version,
     issuerAndSerialNumber IssuerAndSerialNumber,
     digestAlgorithm DigestAlgorithmIdentifier,
     authenticatedAttributes
       [0] IMPLICIT Attributes OPTIONAL,
     digestEncryptionAlgorithm
       DigestEncryptionAlgorithmIdentifier,
     encryptedDigest EncryptedDigest,
     unauthenticatedAttributes
       [1] IMPLICIT Attributes OPTIONAL }

File: src/java.base/share/classes/sun/security/pkcs/SignerInfo.java [hunk 10]
```diff
@@ -162,7 +171,7 @@ public SignerInfo(DerInputStream derin, boolean oldStyle)
         } else {
             // check if set of auth attributes (implicit tag) is provided
             // (auth attributes are OPTIONAL)
-            if ((byte)(derin.peekByte()) == (byte)0xA0) {
+            if (derin.seeOptionalContextSpecific(0)) {
```
Comment (valeriepeng) [inline src/java.base/share/classes/sun/security/pkcs/SignerInfo.java:174]: Again, new call didn't check value at 6th bit.

File: src/java.base/share/classes/sun/security/pkcs/SignerInfo.java [hunk 11]
```diff
@@ -184,8 +193,7 @@ public SignerInfo(DerInputStream derin, boolean oldStyle)
         } else {
             // check if set of unauth attributes (implicit tag) is provided
             // (unauth attributes are OPTIONAL)
-            if (derin.available() != 0
-                && (byte)(derin.peekByte()) == (byte)0xA1) {
+            if (derin.seeOptionalContextSpecific(1)) {
                 unauthenticatedAttributes =
                     new PKCS9Attributes(derin, true);// ignore unsupported attrs
             }
```
Comment (valeriepeng) [inline src/java.base/share/classes/sun/security/pkcs/SignerInfo.java:199]: Later, there is extra bytes under // All done comment, changing that to use atEnd() call? Somehow I cannot add comment to that line? It's not close enough to the lines that you changed.

File: src/java.base/share/classes/sun/security/pkcs/SignerInfo.java (hunk 9 above)
Comment (wangweij) [reply src/java.base/share/classes/sun/security/pkcs/SignerInfo.java:140]: OK.

File: src/java.base/share/classes/sun/security/pkcs/SignerInfo.java (hunk 10 above)
Comment (wangweij) [reply src/java.base/share/classes/sun/security/pkcs/SignerInfo.java:174]: See my previous answer.

File: src/java.base/share/classes/sun/security/pkcs/SignerInfo.java (hunk 11 above)
Comment (wangweij) [reply src/java.base/share/classes/sun/security/pkcs/SignerInfo.java:199]: Good idea.

File: src/java.base/share/classes/sun/security/timestamp/TSResponse.java [hunk 12]
```diff
@@ -332,17 +332,14 @@ private void parse(byte[] tsReply) throws IOException {
...
-                    }
+        if (statusInfo.data.seeOptional(DerValue.tag_SequenceOf)) {
+            DerValue[] strings = statusInfo.data.getSequence(1);
+            statusString = new String[strings.length];
+            for (int i = 0; i < strings.length; i++) {
+                statusString[i] = strings[i].getUTF8String();
+                if (debug != null) {
+                    debug.println("timestamp response: statusString=" +
+                            statusString[i]);
                 }
```
Comment (valeriepeng) [inline src/java.base/share/classes/sun/security/timestamp/TSResponse.java:343]: This is for line 352/349 (see below), would there be extra bytes left after reading the failureInfo bitstring?

File: src/java.base/share/classes/sun/security/tools/keytool/Main.java [hunk 13]
```diff
@@ -1557,7 +1557,11 @@ private void doGenCRL(PrintStream out)
                 badCerts[i] = new X509CRLEntryImpl(new BigInteger(ids.get(i)), firstDate);
             }
         }
-        X509CRLImpl crl = new X509CRLImpl(owner, firstDate, lastDate, badCerts);
+        CRLExtensions crlExts = new CRLExtensions();
+        AuthorityKeyIdentifierExtension ext = new AuthorityKeyIdentifierExtension(
+                new KeyIdentifier(signerCert.getPublicKey()), null, null);
+        crlExts.set(ext.getId(), ext);
```
Comment (valeriepeng) [inline src/java.base/share/classes/sun/security/tools/keytool/Main.java:1563]: Is this for some other bug fix? Seems unrelated to the DER changes...

File: src/java.base/share/classes/sun/security/util/DerValue.java [hunk 14]
```diff
@@ -610,6 +626,17 @@ else if (tag == tag_GeneralString)
             return null;
     }
 
+    public DerValue[] getSubs(byte type) throws IOException {
```
Comment (valeriepeng) [inline src/java.base/share/classes/sun/security/util/DerValue.java:629]: Add comments? It's not very clear what getSubs(...) does just by the method name... It seems to parse and return all DerValue objs from the underlying data bytes?

File: src/java.base/share/classes/sun/security/x509/AuthorityKeyIdentifierExtension.java [hunk 15]
```diff
@@ -155,41 +155,25 @@ public AuthorityKeyIdentifierExtension(Boolean critical, Object value)
                                   "AuthorityKeyIdentifierExtension.");
         }
 
-        // Note that all the fields in AuthorityKeyIdentifier are defined as
-        // being OPTIONAL, i.e., there could be an empty SEQUENCE, resulting
-        // in val.data being null.
```
Comment (valeriepeng) [inline src/java.base/share/classes/sun/security/x509/AuthorityKeyIdentifierExtension.java:160]: Is this comment not valid anymore? Don't we need to check for (val.data != null) in the new code?

File: src/java.base/share/classes/sun/security/x509/AuthorityKeyIdentifierExtension.java [hunk 16]
```diff
@@ -155,41 +155,25 @@ public AuthorityKeyIdentifierExtension(Boolean critical, Object value)
...
-            // The checks only make sense if we assume implicit tagging,
-            // with explicit tagging the form is always constructed.
-            if (opt.isContextSpecific(TAG_ID) && !opt.isConstructed()) {
-                if (id != null)
-                    throw new IOException("Duplicate KeyIdentifier in " +
-                                          "AuthorityKeyIdentifier.");
-                opt.resetTag(DerValue.tag_OctetString);
-                id = new KeyIdentifier(opt);
+        var v = val.data.getOptionalImplicitContextSpecific(
+                TAG_ID, DerValue.tag_OctetString);
```
Comment (valeriepeng) [inline src/java.base/share/classes/sun/security/x509/AuthorityKeyIdentifierExtension.java:159]: Again, the old call has checks for 6th bit, constructed or primitive... New call should also check it.

File: src/java.base/share/classes/sun/security/x509/BasicConstraintsExtension.java [hunk 17]
```diff
@@ -119,85 +119,74 @@ public BasicConstraintsExtension(Boolean critical, boolean ca, int len)
...
-     }
+        if (val.data.seeOptional(DerValue.tag_Boolean)) {
+            this.ca = val.data.getBoolean();
+            if (!this.ca) {
+                throw new IOException("Default cA encoded in BasicConstraints");
+            }
+        }
+        if (this.ca) {
+            if (val.data.seeOptional(DerValue.tag_Integer)) {
+                this.pathLen = val.data.getInteger();
```
Comment (valeriepeng) [inline src/java.base/share/classes/sun/security/x509/BasicConstraintsExtension.java:141]: check pathLen >= 0?

File: src/java.base/share/classes/sun/security/x509/BasicConstraintsExtension.java [hunk 18]
```diff
@@ -119,85 +119,74 @@ public BasicConstraintsExtension(Boolean critical, boolean ca, int len)
...
+     */
+    public void encode(OutputStream out) throws IOException {
+        DerOutputStream tmp = new DerOutputStream();
+        if (extensionValue == null) {
+            this.extensionId = PKIXExtensions.BasicConstraints_Id;
+            if (ca) {
+                critical = true;
+            } else {
+                critical = false;
+            }
```
Comment (valeriepeng) [inline src/java.base/share/classes/sun/security/x509/BasicConstraintsExtension.java:183]: Not relevant to your changes, but it seems strange that the 'critical' value is set solely based on 'ca'. There is a constructor which takes a boolean argument for 'critical'. What is the point of doing that given the "if (ca)" block here?

File: src/java.base/share/classes/sun/security/x509/BasicConstraintsExtension.java (hunk 17 above)
Comment (wangweij) [reply src/java.base/share/classes/sun/security/x509/BasicConstraintsExtension.java:141]: Maybe in another fix? Or do you suggest we can make some small convenient fixes?

File: src/java.base/share/classes/sun/security/x509/BasicConstraintsExtension.java (hunk 18 above)
Comment (wangweij) [reply src/java.base/share/classes/sun/security/x509/BasicConstraintsExtension.java:183]: In fact, I don't know how useful this extension is if ca is false. Maybe the non-critical false flag can be comprehended as "I told you I am not a CA, but if you insist to treat me like one, you can pretend you don't understand my words". How about we leave this unchanged?
